test/upgrade-matrix/mzcompose.py

```diff
@@ -0,0 +1,326 @@
+        default_timeout="300s",
+        no_reset=True,
+        seed=1,
+        entrypoint_extra=[
+            "--var=replicas=1",
+            f"--var=default-replica-size={Materialized.Size.DEFAULT_SIZE}-{Materialized.Size.DEFAULT_SIZE}",
+            f"--var=default-storage-size={Materialized.Size.DEFAULT_SIZE}",
+        ],
+    ),
+]
+
+
+def setup(c: Composition) -> None:
+    c.up("testdrive", persistent=True)
+    c.up("cockroach", "redpanda", "postgres", "debezium")
+
+
+def teardown(c: Composition) -> None:
+    c.down(destroy_volumes=True)
+
+
+def workflow_default(c: Composition, parser: WorkflowArgumentParser) -> None:
+    """Test sequences of upgrades across the entire possible matrix of upgrade scenarios.
+
+    Thie framework works by building a graph that contains the Mz versions and Checks actions
+    to perform as nodes, as well as the allowed upgrade paths and other required dependencies
+    (as edges). All valid simple paths within the graph are upgrade scenarios that need to be
+    exercised and we pick a subset of them to run.
+    """
+
+    c.silent = True
+    parser.add_argument(
+        "--check",
+        metavar="CHECK",
+        type=str,
+        action="append",
+        help="Check(s) to run.",
+        default=["PgCdc", "SinkUpsert"],
+    )
+
+    parser.add_argument(
+        "--dry-run",
+        action="store_true",
+        help="Just print the upgrade paths that will be tested.",
+    )
+
+    parser.add_argument(
+        "--min-version",
+        metavar="VERSION",
+        type=MzVersion.parse,
+        default=MzVersion.parse("0.39.0"),
+        help="Minimum Mz version to involve.",
+    )
+
+    parser.add_argument(
+        "--max-version-gap",
+        type=int,
+        default=1,
+        help="Maximum number of versions to skip",
+    )
+
+    parser.add_argument(
+        "--num-scenarios",
+        type=int,
+        default=50,
+        help="Number of upgrade scenarios to test",
+    )
+
+    parser.add_argument("--seed", metavar="SEED", type=str, default=str(time.time()))
+
+    parser.add_argument("--scenario-id", type=int, default=None)
+
+    args = parser.parse_args()
+
+    print(f"--- Random seed is {args.seed}")
+    random.seed(args.seed)
+
+    checks = (
+        [globals()[check] for check in args.check]
+        if args.check
+        else Check.__subclasses__()
+    )
+
+    print(f"--- Checks to use: {checks}")
+
+    executor = MzcomposeExecutor(composition=c)
+
+    versions = list(reversed([v for v in all_versions if v >= args.min_version]))
+    print(
+        "--- Testing upgrade scenarios involving the following versions: "
+        + " ".join([str(v) for v in versions])
+    )
+
+    for id, upgrade_scenario in enumerate(
+        get_upgrade_scenarios(
+            versions=versions,
+            max_version_gap=args.max_version_gap,
+            num_scenarios=args.num_scenarios,
+        )
+    ):
+        if args.scenario_id and id != args.scenario_id:
```

Comment:
```suggestion
        if not args.scenario_id is None and id != args.scenario_id:
```
Otherwise scenario_id 0 might not work if I'm reading this correctly.

Reply:
Thanks, fixed.